_posts/blog/2016-10-28-durability-performance.md

```diff
@@ -0,0 +1,136 @@
+   * Reports success back to client.
+
+The sync/flush step above moves data written to the WAL from memory to disk.
+Write ahead logs are stored in HDFS. HDFS supports two ways of forcing data to
+disk for an open file : `hsync` and `hflush`.  
+
+## HDFS Sync/Flush Details
+
+When `hflush` is called on a WAL, it does not ensure data is on disk.  It only
+ensure that data is in OS buffers on each datanode and on its way to disk.  As a
+result calls to `hflush` are very fast.  If a WAL is replicated to 3 data nodes
+then data may be lost if all three machines reboot.  If the datanode process
+dies, thats ok because it flushed to OS.  The machines have to reboot for data
+loss to occur.
+
+In order to avoid data loss in the event of reboot, `hsync` can be called.  This
+will ensure data is written to disk on all datanodes before returning.  When
+using `hsync` for the WAL, if Accumulo reports success to a user it means the
+data is on disk.  However `hsync` is much slower than `hflush` and the way it's
+implemented exacerbates the problem.  For example `hflush` make take 1ms and
+`hsync` may take 50ms.  This difference will impact writes to Accumulo and can
+be mitigated in some situations with larger buffers in Accumulo.
+
+HDFS keeps checksum data internally by default.  Datanodes store checksum data
+in a separate file in the local filesystem.  This means when `hsync` is called
+on a WAL, two files must be synced on each datanode.  Syncing two files doubles
+the time. To make matters even worse, when the two files are synced the local
+filesystem metadata is also synced.  Depending on the local filesystem and its
+configuration, syncing the metadata may or may not take time.  In the worst
+case, we need to wait for four sync operations at the local filesystem level on
+each datanode. One thing I am not sure about, is if these sync operations occur
+in parallel on the replicas on different datanodes.  Lets hope they occur in
+parallel.  The following pointers show where sync occurs in the datanode code.
+
+ * [BlockReceiver.flushOrSync()][fos] calls [ReplicaOutputStreams.syncDataOut()][ros1] and [ReplicaOutputStreams.syncChecksumOut()][ros2] when `isSync` is true.
+ * The methods in ReplicaOutputStreams call [FileChannel.force(true)][fcf] which
+   synchronously flushes data and filesystem metadata.
+
+If files were preallocated (this would avoid syncing local filesystem metadata)
+and checksums were stored in-line, then 1 sync could be done instead of 4.  
+
+## Configuring WAL flush/sync in Accumulo 1.6
+
+Accumulo 1.6.0 only supported `hsync` and this caused [performance
+problems][160_RN_WAL].  In order to offer better performance, the option to
+configure `hflush` was [added in 1.6.1][161_RN_WAL].  The
+[tserver.wal.sync.method][16_UM_SM] configuration option was added to support
+this feature.  This was a tablet server wide option that applied to everything
+written to any table.   
+
+## Group Commit
```

Comment:
I had it like that initially.  Then I realized that the information about group commit was useful for understanding the 1.7 section, so I moved it.  Does this change your opinion?